In [1]:
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Intro to Generating and Executing Python Code with Gemini 2.5 Flash

<table align="left">
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo"><br> Open in Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https:%2F%2Fraw.githubusercontent.com%2FGoogleCloudPlatform%2Fgenerative-ai%2Fmain%2Fgemini%2Fcode-execution%2Fintro_code_execution.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo"><br> Open in Colab Enterprise
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/GoogleCloudPlatform/generative-ai/main/gemini/code-execution/intro_code_execution.ipynb">
      <img src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo"><br> Open in Vertex AI Workbench
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo"><br> View on GitHub
    </a>
  </td>
</table>

<div style="clear: both;"></div>

<b>Share to:</b>

<a href="https://www.linkedin.com/sharing/share-offsite/?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="LinkedIn logo">
</a>

<a href="https://bsky.app/intent/compose?text=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="Bluesky logo">
</a>

<a href="https://twitter.com/intent/tweet?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X logo">
</a>

<a href="https://reddit.com/submit?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit logo">
</a>

<a href="https://www.facebook.com/sharer/sharer.php?u=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/51/Facebook_f_logo_%282019%29.svg" alt="Facebook logo">
</a>

| | |
|-|-|
| Author(s) |  [Kristopher Overholt](https://github.com/koverholt/) |

## Overview

This notebook introduces the code execution capabilities of the [Gemini 2.5 Flash model](https://cloud.google.com/vertex-ai/generative-ai/docs/models), a new multimodal generative AI model from Google [DeepMind](https://deepmind.google/). Gemini 2.5 Flash offers improvements in speed, quality, and advanced reasoning capabilities including enhanced understanding, coding, and instruction following.

## Code Execution

A key feature of this model is [code execution](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution), which is the ability to generate and execute Python code directly within the API. If you want the API to generate and run Python code and return the results, you can use code execution as demonstrated in this notebook.

This code execution capability enables the model to generate code, execute and observe the results, correct the code if needed, and learn iteratively from the results until it produces a final output. This is particularly useful for applications that involve code-based reasoning such as solving mathematical equations or processing text.

## Objectives

In this tutorial, you will learn how to generate and execute code using the Gemini API in Vertex AI and the Google Gen AI SDK for Python with the Gemini 2.5 Flash model.

You will complete the following tasks:

- Generating and running sample Python code from text prompts
- Exploring data using code execution in multi-turn chats
- Using code execution in streaming sessions

## Getting started

### Install Google Gen AI SDK for Python


In [2]:
%pip install --upgrade --quiet google-genai

Note: you may need to restart the kernel to use updated packages.


### Import libraries

In [3]:

from IPython.display import Markdown, display
from google import genai
from google.genai.types import GenerateContentConfig, Tool, ToolCodeExecution

### Connect to a generative AI API service

Google Gen AI APIs and models including Gemini are available in the following two API services:

- [Google AI for Developers](https://ai.google.dev/gemini-api/docs): Experiment, prototype, and deploy small projects.
- [Vertex AI](https://cloud.google.com/vertex-ai/generative-ai/docs): Build enterprise-ready projects on Google Cloud.
The Google Gen AI SDK provides a unified interface to these two API services.

This notebook shows how to use the Google Gen AI SDK with the Gemini API in Vertex AI.

### Set Google Cloud project information and create client

To get started using Vertex AI, you must have an existing Google Cloud project and [enable the Vertex AI API](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com).

Learn more about [setting up a project and a development environment](https://cloud.google.com/vertex-ai/docs/start/cloud-environment).

In [4]:
PROJECT_ID = "qwiklabs-gcp-01-26236439ed1e"
LOCATION = "global"
client = genai.Client(vertexai=True, project=PROJECT_ID, location=LOCATION)


## Working with code execution in Gemini 2.5 Flash

### Load the Gemini model

The following code loads the Gemini 2.5 Flash model. You can learn about all Gemini models on Vertex AI by visiting the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models):

In [5]:
MODEL_ID = "gemini-2.5-flash"  # @param {type: "string"}

### Define the code execution tool

The following code initializes the code execution tool by passing `code_execution` in a `Tool` definition.

Later we'll register this tool with the model that it can use to generate and run Python code:

In [6]:
code_execution_tool = Tool(code_execution=ToolCodeExecution())

### Generate and execute code

The following code sends a prompt to the Gemini model, asking it to generate and execute Python code to calculate the sum of the first 50 prime numbers. The code execution tool is passed in so the model can generate and run the code:

In [7]:
PROMPT = """What is the sum of the first 50 prime numbers?
Generate and run code for the calculation."""

response = client.models.generate_content(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

### View the generated code

The following code iterates through the response and displays any generated Python code by checking for `part.executable_code` in the response parts:

In [8]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )


```py
import math

def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

prime_numbers = []
num = 2
while len(prime_numbers) < 50:
    if is_prime(num):
        prime_numbers.append(num)
    num += 1

sum_of_primes = sum(prime_numbers)
print(f'{prime_numbers=}')
print(f'{sum_of_primes=}')
```


### View the code execution results

The following code iterates through the response and displays the execution result and outcome by checking for `part.code_execution_result` in the response parts:

In [14]:
for part in response.candidates[0].content.parts:
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)

`prime_numbers=[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229]
sum_of_primes=5117
`


Outcome: Outcome.OUTCOME_OK


Great! Now you have the answer (`5117`) as well as the generated (and verified via execution!) Python code.

At this point in your application, you would save the output code, result, or outcome and display it to the end-user or use it downstream in your application.

### Code execution in a chat session

This section shows how to use code execution in an interactive chat with history using the Gemini API.

You can use `client.chats.create` to create a chat session and passes in the code execution tool, enabling the model to generate and run code:

In [21]:
chat = client.chats.create(
    model=MODEL_ID,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
        max_output_tokens=10240,
    ),
)

You'll start the chat by asking the model to generate sample time series data with noise and then output a sample of 10 data points:

In [22]:
PROMPT = """Create sample time series data of temperature vs. time in a test furnace.
Add noise to the data. Output a sample of 10 data points from the time series data."""

response = chat.send_message(PROMPT)

Now you can iterate through the response to display any generated Python code and execution results by checking for `part.executable_code` and `part.code_execution_result` in the response parts:

In [23]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py
import numpy as np

# --- Parameters for the furnace temperature profile ---
T_start = 25.0  # Initial temperature (e.g., room temperature in Celsius)
T_setpoint = 500.0 # Target furnace temperature
heating_rate = 25.0 # Degrees Celsius per unit of time (e.g., per minute)
noise_amplitude = 3.0 # Standard deviation of the noise in Celsius

# --- Time parameters ---
total_duration = 30.0 # Total duration of the simulation (e.g., in minutes)
sampling_interval = 0.5 # Time step between measurements (e.g., 30 seconds)

# Generate time points
time = np.arange(0, total_duration + sampling_interval, sampling_interval)

# Calculate base temperature profile
# The temperature rises linearly until it reaches the setpoint, then stays constant.
base_temperature = np.minimum(T_setpoint, T_start + heating_rate * time)

# Add Gaussian noise to the base temperature
noise = np.random.normal(0, noise_amplitude, len(time))
noisy_temperature = base_temperature + noise

# --- Select 10 sample data points ---
# To get a representative sample, we'll pick 10 evenly spaced points.
# If there are fewer than 10 points, we'll just take all of them.
num_total_points = len(time)
if num_total_points >= 10:
    indices = np.linspace(0, num_total_points - 1, 10, dtype=int)
else:
    indices = np.arange(num_total_points)

sample_time = time[indices]
sample_temperature = noisy_temperature[indices]

print("Sample Time Series Data (Time vs. Temperature with Noise):")
print("-------------------------------------------------------")
for i in range(len(sample_time)):
    print(f"Time: {sample_time[i]:.1f} min, Temperature: {sample_temperature[i]:.2f} °C")

```


`Sample Time Series Data (Time vs. Temperature with Noise):
-------------------------------------------------------
Time: 0.0 min, Temperature: 22.27 °C
Time: 3.0 min, Temperature: 94.90 °C
Time: 6.5 min, Temperature: 186.45 °C
Time: 10.0 min, Temperature: 272.90 °C
Time: 13.0 min, Temperature: 351.89 °C
Time: 16.5 min, Temperature: 440.24 °C
Time: 20.0 min, Temperature: 499.32 °C
Time: 23.0 min, Temperature: 498.27 °C
Time: 26.5 min, Temperature: 502.08 °C
Time: 30.0 min, Temperature: 503.36 °C
`


Outcome: Outcome.OUTCOME_OK


Now you can ask the model to add a smoothed data series to the time series data:

In [24]:
PROMPT = "Now add a data series that smooths the sample data."

response = chat.send_message(PROMPT)

And then display the generated Python code and execution results:

In [25]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py
import numpy as np

# --- Parameters for the furnace temperature profile (same as before) ---
T_start = 25.0
T_setpoint = 500.0
heating_rate = 25.0
noise_amplitude = 3.0

# --- Time parameters (same as before) ---
total_duration = 30.0
sampling_interval = 0.5

# Generate time points
time = np.arange(0, total_duration + sampling_interval, sampling_interval)

# Calculate base temperature profile
base_temperature = np.minimum(T_setpoint, T_start + heating_rate * time)

# Add Gaussian noise to the base temperature
noise = np.random.normal(0, noise_amplitude, len(time))
noisy_temperature = base_temperature + noise

# --- Select 10 sample data points (same as before) ---
num_total_points = len(time)
if num_total_points >= 10:
    indices = np.linspace(0, num_total_points - 1, 10, dtype=int)
else:
    indices = np.arange(num_total_points)

sample_time = time[indices]
sample_temperature = noisy_temperature[indices]

# --- Add Smoothing ---
# Define the window size for the moving average
window_size = 3

# Apply a moving average filter
# np.convolve is used for convolution. We create a window of ones and normalize it.
# 'valid' mode would shorten the array, 'same' mode pads and keeps the same length.
# For 'same' mode, padding is applied to the ends.
# For a simple moving average, we can manually handle edges or use a function like pandas.rolling.mean
# or scipy.ndimage.uniform_filter1d.
# For this small sample, a manual approach or a simple convolution with 'same' mode is fine.

# Manual moving average for 'same' mode
smoothed_temperature = np.zeros_like(sample_temperature)
for i in range(len(sample_temperature)):
    start_idx = max(0, i - window_size // 2)
    end_idx = min(len(sample_temperature), i + window_size // 2 + 1)
    smoothed_temperature[i] = np.mean(sample_temperature[start_idx:end_idx])


print("Sample Time Series Data (Time vs. Temperature with Noise and Smoothed):")
print("--------------------------------------------------------------------")
print(f"{'Time (min)':<12} | {'Noisy Temp (°C)':<18} | {'Smoothed Temp (°C)':<18}")
print("--------------------------------------------------------------------")
for i in range(len(sample_time)):
    print(f"{sample_time[i]:<12.1f} | {sample_temperature[i]:<18.2f} | {smoothed_temperature[i]:<18.2f}")

```


`Sample Time Series Data (Time vs. Temperature with Noise and Smoothed):
--------------------------------------------------------------------
Time (min)   | Noisy Temp (°C)    | Smoothed Temp (°C)
--------------------------------------------------------------------
0.0          | 23.04              | 64.08             
3.0          | 105.13             | 105.91            
6.5          | 189.58             | 188.00            
10.0         | 269.31             | 268.38            
13.0         | 346.25             | 352.02            
16.5         | 440.51             | 428.90            
20.0         | 499.95             | 482.41            
23.0         | 506.77             | 501.99            
26.5         | 499.27             | 501.06            
30.0         | 497.14             | 498.20            
`


Outcome: Outcome.OUTCOME_OK


Finally, you can ask the model to generate descriptive statistics for the time series data:

In [26]:
PROMPT = "Now generate and output descriptive statistics on the time series data."

response = chat.send_message(PROMPT)

And then display the generated Python code and execution results:

In [27]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py
import numpy as np

# --- Parameters for the furnace temperature profile (same as before) ---
T_start = 25.0
T_setpoint = 500.0
heating_rate = 25.0
noise_amplitude = 3.0

# --- Time parameters (same as before) ---
total_duration = 30.0
sampling_interval = 0.5

# Generate time points
time = np.arange(0, total_duration + sampling_interval, sampling_interval)

# Calculate base temperature profile
base_temperature = np.minimum(T_setpoint, T_start + heating_rate * time)

# Add Gaussian noise to the base temperature
noise = np.random.normal(0, noise_amplitude, len(time))
noisy_temperature = base_temperature + noise

# --- Select 10 sample data points (same as before) ---
num_total_points = len(time)
if num_total_points >= 10:
    indices = np.linspace(0, num_total_points - 1, 10, dtype=int)
else:
    indices = np.arange(num_total_points)

sample_time = time[indices]
sample_temperature = noisy_temperature[indices]

# --- Add Smoothing (same as before) ---
window_size = 3
smoothed_temperature = np.zeros_like(sample_temperature)
for i in range(len(sample_temperature)):
    start_idx = max(0, i - window_size // 2)
    end_idx = min(len(sample_temperature), i + window_size // 2 + 1)
    smoothed_temperature[i] = np.mean(sample_temperature[start_idx:end_idx])

# --- Generate Descriptive Statistics ---
def get_descriptive_stats(data, name):
    stats = {
        "Series": name,
        "Count": len(data),
        "Mean": np.mean(data),
        "Median": np.median(data),
        "Std Dev": np.std(data),
        "Min": np.min(data),
        "Max": np.max(data),
        "Range": np.max(data) - np.min(data)
    }
    return stats

stats_noisy = get_descriptive_stats(sample_temperature, "Noisy Temperature")
stats_smoothed = get_descriptive_stats(smoothed_temperature, "Smoothed Temperature")

print("Descriptive Statistics for Temperature Data:")
print("--------------------------------------------")

# Print header
print(f"{'Statistic':<15} | {stats_noisy['Series']:<20} | {stats_smoothed['Series']:<20}")
print("-" * 60)

# Print values
for key in ["Count", "Mean", "Median", "Std Dev", "Min", "Max", "Range"]:
    print(f"{key:<15} | {stats_noisy[key]:<20.2f} | {stats_smoothed[key]:<20.2f}")

```


`Descriptive Statistics for Temperature Data:
--------------------------------------------
Statistic       | Noisy Temperature    | Smoothed Temperature
------------------------------------------------------------
Count           | 10.00                | 10.00               
Mean            | 338.68               | 339.86              
Median          | 397.58               | 393.84              
Std Dev         | 172.37               | 162.93              
Min             | 24.63                | 61.89               
Max             | 502.91               | 501.07              
Range           | 478.28               | 439.18              
`


Outcome: Outcome.OUTCOME_OK


This chat example demonstrates how you can use the Gemini API with code execution as a powerful tool for exploratory data analysis and more. Go forth and adapt this approach to your own projects and use cases!

### Code execution in a streaming session

You can also use the code execution functionality with streaming output from the Gemini API.

The following code demonstrates how the Gemini API can generate and execute code while streaming the results:

In [28]:
PROMPT = """Generate a list of 20 random names, then create a new list with just
the names containing the letter 'a', then output the number of names that
contain 'a' and finally show me that new list."""

for chunk in client.models.generate_content_stream(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
):
    if chunk.candidates and chunk.candidates[0].content:
        if chunk.candidates[0].content.parts is not None:
            for part in chunk.candidates[0].content.parts:
                if part.text:
                    display(Markdown("#### Natural language stream"))
                    display(Markdown(part.text))
                    display(Markdown("---"))
                if part.executable_code:
                    display(Markdown("#### Code stream"))
                    display(
                        Markdown(
                            f"""
    ```py
    {part.executable_code.code}
    ```
    """
                        )
                    )
                    display(Markdown("---"))
                if part.code_execution_result:
                    display(Markdown("#### Code result"))
                    display(
                        Markdown(
                            f"""
    ```
    {part.code_execution_result.output}
    ```
    """
                        )
                    )
                    display(Markdown("---"))

#### Code stream


    ```py
    import random

# Step 1: Create a list of potential names
potential_names = [
    "Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Grace", "Heidi", "Ivan", "Judy",
    "Kevin", "Linda", "Mike", "Nancy", "Oscar", "Pamela", "Quinn", "Rachel", "Steve", "Tina",
    "Uma", "Victor", "Wendy", "Xavier", "Yara", "Zoe", "Adam", "Anna", "Brian", "Carla",
    "Daniel", "Diana", "Ethan", "Hannah", "Isabella", "Jacob", "Karen", "Laura", "Mark",
    "Maria", "Nathan", "Olivia", "Patrick", "Paula", "Robert", "Sarah", "Thomas", "Victoria",
    "William", "Samantha", "Alexander", "Amanda", "Andrew", "Angela", "Anthony", "Barbara",
    "Benjamin", "Brenda", "Catherine", "Charles", "Christine", "Christopher", "Cynthia",
    "Deborah", "Dennis", "Donald", "Donna", "Dorothy", "Edward", "Elizabeth", "Emily",
    "Eric", "Evelyn", "Florence", "Frances", "Gary", "George", "Georgia", "Gerald",
    "Gloria", "Gregory", "Harold", "Helen", "Henry", "Jacqueline", "James", "Janet",
    "Jason", "Jeffrey", "Jennifer", "Jerry", "Jessica", "Joan", "Joe", "John", "Jonathan",
    "Jose", "Joseph", "Joshua", "Joyce", "Juan", "Judith", "Julia", "Julie", "Justin",
    "Kathleen", "Kelly", "Kenneth", "Kimberly", "Larry", "Lawrence", "Lisa", "Louis",
    "Louise", "Margaret", "Marilyn", "Martha", "Mary", "Matthew", "Megan", "Melissa",
    "Michael", "Michelle", "Nancy", "Nicholas", "Nicole", "Patricia", "Paul", "Peter",
    "Philip", "Raymond", "Rebecca", "Richard", "Ronald", "Rose", "Roy", "Russell",
    "Sandra", "Scott", "Sharon", "Shirley", "Stephen", "Stephanie", "Steven", "Susan",
    "Teresa", "Terry", "Theresa", "Timothy", "Virginia", "Walter", "Wayne", "William"
]

# Step 2: Generate 20 random names
random_names = random.sample(potential_names, 20)
print(f"Original list of 20 random names:\n{random_names}\n")

# Step 3 & 4: Initialize an empty list for names containing 'a' and iterate
names_with_a = []
for name in random_names:
    # Step 5: Check if the name contains 'a' (case-insensitive)
    if 'a' in name.lower():
        # Step 6: If it does, add it to the new list
        names_with_a.append(name)

# Step 7: Output the number of names that contain 'a'
print(f"Number of names containing 'a': {len(names_with_a)}\n")

# Step 8: Show the new list
print(f"List of names containing 'a':\n{names_with_a}")
    ```
    

---

#### Code result


    ```
    Original list of 20 random names:
['Eric', 'Larry', 'Rose', 'David', 'Jason', 'Hannah', 'Rebecca', 'Scott', 'Gloria', 'Emily', 'Marilyn', 'Nathan', 'Dennis', 'Daniel', 'Henry', 'Theresa', 'Donald', 'William', 'Russell', 'Eve']

Number of names containing 'a': 12

List of names containing 'a':
['Larry', 'David', 'Jason', 'Hannah', 'Rebecca', 'Gloria', 'Marilyn', 'Nathan', 'Daniel', 'Theresa', 'Donald', 'William']

    ```
    

---

#### Natural language stream

Here

---

#### Natural language stream

 is the list of 20 random names, followed by a new list containing only the

---

#### Natural language stream

 names with the letter 'a', the count of those names, and finally the list itself.



---

#### Natural language stream

Original list of 20 random names:
['Eric', 'Larry', 'Rose', 'David', 'Jason', 'Hannah', 'Rebecca', 'Scott', 'Gloria',

---

#### Natural language stream

 'Emily', 'Marilyn', 'Nathan', 'Dennis', 'Daniel', 'Henry', 'Theresa', 'Donald', 'William', 'Russell', 'Eve']

---

#### Natural language stream



Number of names containing 'a': 12

List of names containing 'a':
['Larry', 'David', 'Jason', 'Hannah', 'Rebecca', 'Gloria', 'Marilyn', 'Nathan', 'Daniel

---

#### Natural language stream

', 'Theresa', 'Donald', 'William']

---

This streaming example demonstrated how the Gemini API can generate, execute code, and provide results within a streaming session.

## Summary

Refer to the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution) for more details about code execution, and in particular, the [recommendations](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution#code-execution-vs-function-calling) regarding differences between code execution and [function calling](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling).

### Next steps

- See the [Google Gen AI SDK reference docs](https://googleapis.github.io/python-genai/)
- Explore other notebooks in the [Google Cloud Generative AI GitHub repository](https://github.com/GoogleCloudPlatform/generative-ai)
- Explore AI models in [Model Garden](https://cloud.google.com/vertex-ai/generative-ai/docs/model-garden/explore-models)